README.md
# Contao Fileinfo Bundle
File infos (meta and copyright) are inherited from parent folder

copyright datas expect the package: heimrichhannot/contao-filecredits


CHANGELOG
v1.1
add: compatibility for Contao 4.9

v1.0.4
fix: text correction

v1.0.3
fix: inherit meta from folder

v1.0.2
fix: save file

v1.0.1
fix: config.php

v1.0.0
Init

src/ContaoManager/Plugin.php
<?php
/**
 *  Copyright Information
 *
 * @copyright: 2018 agentur fipps e.K.
 * @author   : Arne Borchert
 * @license  : LGPL 3.0+
 */

namespace Fipps\FileinfoBundle\ContaoManager;

use Fipps\FileinfoBundle\FippsFileinfoBundle;
use Contao\CoreBundle\ContaoCoreBundle;
use Contao\ManagerPlugin\Bundle\Config\BundleConfig;
use Contao\ManagerPlugin\Bundle\BundlePluginInterface;
use Contao\ManagerPlugin\Bundle\Parser\ParserInterface;

/**
 * Plugin for the Contao Manager.
 */
class Plugin implements BundlePluginInterface
{
    /**
     * {@inheritdoc}
     */
    public function getBundles(ParserInterface $parser)
    {
        return [
            BundleConfig::create(FippsFileinfoBundle::class)->setLoadAfter(
                [
                    ContaoCoreBundle::class,
                    'filecredits'
                ]
            )
        ];
    }

    /**
     * {@inheritdoc}
     */
    public function registerContainerConfiguration(LoaderInterface $loader, array $config)
    {
        $file = '@FippsFileinfoBundle/Resources/config/config.yml';
        
        $loader->load($file);
    }

    /**
     * {@inheritdoc}
     */
    public function getRouteCollection(LoaderResolverInterface $resolver, KernelInterface $kernel)
    {
        $file = '@FippsFileinfoBundle/Resources/config/routing.yml';

        return $resolver->resolve($file)->load($file);
    }

    /**
     * {@inheritdoc}
     */
    public function getPackageDependencies()
    {
        return ['heimrichhannot/contao-filecredits'];
    }
}

src/Resources/contao/dca/tl_files.php
<?php
/**
 *  Copyright Information
 *  @copyright: 2019 agentur fipps e.K.
 *  @author   : Arne Borchert <arne.borchert@fipps.de>
 *  @license  : LGPL 3.0+
 */

$GLOBALS['TL_DCA']['tl_files']['config']['onload_callback'][] = array(\Fipps\FileinfoBundle\DataContainer\FileCallback::class, 'addMetaToFolder');

$GLOBALS['TL_DCA']['tl_files']['fields']['meta']['load_callback'][] = array(\Fipps\FileinfoBundle\DataContainer\FileCallback::class, 'loadMetaCallback');
$GLOBALS['TL_DCA']['tl_files']['fields']['meta']['save_callback'][] = array(\Fipps\FileinfoBundle\DataContainer\FileCallback::class, 'saveMetaCallback');

if (is_array($GLOBALS['TL_DCA']['tl_files']['fields']['copyright'])) {
    $GLOBALS['TL_DCA']['tl_files']['fields']['copyright']['save_callback'][] = array(\Fipps\FileinfoBundle\DataContainer\FileCallback::class, 'saveCopyrightCallback');
}

src/Resources/contao/config/config.php
<?php
/**
 *
 *   Copyright Information
 *   @copyright: 2018 agentur fipps e.K., 2016-2018 Richard Henkenjohann
 *   @author   : Arne Borchert, Richard Henkenjohann <richardhenkenjohann@googlemail.com>
 *   @license  : LGPL 3.0+
 */


/**
 * -------------------------------------------------------------------------
 * BACK END MODULES
 * -------------------------------------------------------------------------
 *
 * Back end modules are stored in a global array called "BE_MOD". Each module
 * has certain properties like an icon, an optional callback function and one
 * or more tables. Each module belongs to a particular group.
 *
 * $GLOBALS['BE_MOD'] = array
 * (
 * 'group_1' => array
 * (
 * 'module_1' => array
 * (
 * 'tables' => array('table_1', 'table_2'),
 * 'key' => array('Class', 'method'),
 * 'callback' => 'ClassName',
 * 'icon' => 'path/to/icon.gif',
 * 'stylesheet' => 'path/to/stylesheet.css',
 * 'javascript' => 'path/to/javascript.js'
 * )
 * )
 * );
 *
 * Use function array_insert() to modify an existing modules array.
 */


/**
 * -------------------------------------------------------------------------
 * FRONT END MODULES
 * -------------------------------------------------------------------------
 *
 * List all front end modules and their class names.
 *
 * $GLOBALS['FE_MOD'] = array
 * (
 * 'group_1' => array
 * (
 * 'module_1' => 'Contentlass',
 * 'module_2' => 'Contentlass'
 * )
 * );
 *
 * Use function array_insert() to modify an existing CTE array.
 */

/**
 * -------------------------------------------------------------------------
 * CONTENT ELEMENTS
 * -------------------------------------------------------------------------
 *
 * List all content elements and their class names.
 *
 * $GLOBALS['TL_CTE'] = array
 * (
 * 'group_1' => array
 * (
 * 'cte_1' => 'Contentlass',
 * 'cte_2' => 'Contentlass'
 * )
 * );
 *
 * Use function array_insert() to modify an existing CTE array.
 */

/**
 * -------------------------------------------------------------------------
 * BACK END FORM FIELDS
 * -------------------------------------------------------------------------
 *
 * List all back end form fields and their class names.
 *
 * $GLOBALS['BE_FFL'] = array
 * (
 * 'input' => 'Class',
 * 'select' => 'Class'
 * );
 *
 * Use function array_insert() to modify an existing FFL array.
 */

/**
 * -------------------------------------------------------------------------
 * FRONT END FORM FIELDS
 * -------------------------------------------------------------------------
 *
 * List all form fields and their class names.
 *
 * $GLOBALS['TL_FFL'] = array
 * (
 * 'input' => Class,
 * 'select' => Class
 * );
 *
 * Use function array_insert() to modify an existing FFL array.
 */

/**
 * -------------------------------------------------------------------------
 * CACHE TABLES
 * -------------------------------------------------------------------------
 *
 * These tables are used to cache data and can be truncated using back end
 * module "clear cache".
 *
 * $GLOBALS['TL_CACHE'] = array
 * (
 * 'table_1',
 * 'table_2'
 * );
 *
 * Use function array_insert() to modify an existing cache array.
 */

/**
 * -------------------------------------------------------------------------
 * HOOKS
 * -------------------------------------------------------------------------
 *
 * Hooking allows you to register one or more callback functions that are
 * called on a particular event in a specific order. Thus, third party
 * extensions can add functionality to the core system without having to
 * modify the source code.
 *
 * $GLOBALS['TL_HOOKS'] = array
 * (
 * 'hook_1' => array
 * (
 * array('Class', 'Method'),
 * array('Class', 'Method')
 * )
 * );
 *
 * Use function array_insert() to modify an existing hooks array.
 */
$GLOBALS['TL_HOOKS']['postUpload'][] = array(\Fipps\FileinfoBundle\Listener\HooksListener::class, 'setFileInfoAfterUpload');


/**
 * -------------------------------------------------------------------------
 * PAGE TYPES
 * -------------------------------------------------------------------------
 *
 * Page types and their corresponding front end controller class.
 *
 * $GLOBALS['TL_PTY'] = array
 * (
 * 'type_1' => 'PageType1',
 * 'type_2' => 'PageType2'
 * );
 *
 * Use function array_insert() to modify an existing page types array.
 */

/**
 * -------------------------------------------------------------------------
 * Models
 * -------------------------------------------------------------------------
 *
 * Register models with seperate namespaces
 *
 * $GLOBALS['TL_MODELS'] = array
 * (
 * 'table' => 'Class',
 * );
 * 
 * Ex.: $GLOBALS['TL_MODELS'][Class::getTable()] = Class::class; 
 * 
 * Use function array_insert() to modify an existing page types array.
 */



src/DependencyInjection/FippsFileinfoExtension.php
<?php
/**
 *  Copyright Information
 *
 * @copyright: 2018 agentur fipps e.K.
 * @author   : Arne Borchert
 * @license  : LGPL 3.0+
 */

namespace Fipps\FileinfoBundle\DependencyInjection;

use Symfony\Component\Config\FileLocator;
use Symfony\Component\DependencyInjection\ContainerBuilder;
use Symfony\Component\HttpKernel\DependencyInjection\Extension;
use Symfony\Component\DependencyInjection\Loader\YamlFileLoader;

/**
 * Adds the bundle services to the container.
 */
class FippsFileinfoExtension extends Extension
{
    /**
     * {@inheritdoc}
     */
    public function load(array $mergedConfig, ContainerBuilder $container)
    {
        $loader = new YamlFileLoader(
            $container, new FileLocator(__DIR__.'/../Resources/config')
        );
        $loader->load('services.yml');
    }
}


src/DataContainer/FileCallback.php
<?php
/**
 *  Copyright Information
 *
 * @copyright: 2019 agentur fipps e.K.
 * @author   : Arne Borchert <arne.borchert@fipps.de>
 * @license  : LGPL 3.0+
 */

namespace Fipps\FileinfoBundle\DataContainer;


use DC_Folder;
use FilesModel;
use DataContainer;
use StringUtil;
use Contao\CoreBundle\DataContainer\PaletteManipulator;
use Contao\System;

class FileCallback
{

    public function addMetaToFolder(DataContainer $dc)
    {
        if (!$dc->id) {
            return;
        }

        $projectDir  = System::getContainer()->getParameter('kernel.project_dir');
        $blnIsFolder = is_dir($projectDir.'/'.$dc->id);

        // Add the meta data when editing folders


        if ($blnIsFolder && version_compare(VERSION, '4.9', '>=')) {
            PaletteManipulator::create()
                ->addField('meta', 'copyright')
                ->applyToPalette('default', $dc->table);
        }
        $GLOBALS['TL_DCA']['tl_files']['fields']['meta']['eval']['tl_class'] = 'clr';

    }

    /**
     * Bugfix
     *
     * @param string         $val
     * @param \DataContainer $dc
     * @return string
     */
    public function loadMetaCallback($val, DataContainer $dc)
    {
        if (is_array(StringUtil::deserialize($val))) {
            return $val;
        }

        return;
    }

    /**
     * @param string         $val
     * @param \DataContainer $dc
     * @return string
     */
    public function saveMetaCallback($val, DataContainer $dc)
    {
        $arrMetaNew = unserialize($val);
        if (!is_array($arrMetaNew)) {
            return serialize([]);
        }
        if ($dc instanceof DC_Folder) {
            $activeRecord = $dc->activeRecord;
            $strOldMeta   = $activeRecord->meta;
            $arrOldMeta   = unserialize($strOldMeta);
            $folder       = FilesModel::findByUuid($activeRecord->uuid);
            $files        = FilesModel::findByPid($folder->uuid);

            if ($files !== null) {
                foreach ($files as $file) {
                    if ($file instanceof DC_Folder) {
                        continue;
                    }
                    // Falls File Meta leer oder dem alten Meta des Folders entspricht
                    if ($file->meta === null || $file->meta == '' || $file->meta == $strOldMeta) {
                        $file->meta = $val;
                        $file->save();
                        continue;
                    }

                    // Check every lang entry
                    $arrFileMeta = unserialize($file->meta);
                    foreach ($arrMetaNew as $lang => $arrMetaContent) {
                        if (isset($arrFileMeta[$lang])) {
                            if ($arrFileMeta[$lang] == $arrOldMeta[$lang]) {
                                $arrFileMeta[$lang] = $arrMetaContent;
                            }
                        } else {
                            $arrFileMeta[$lang] = $arrMetaContent;
                        }
                    }

                    $file->meta = serialize($arrFileMeta);
                    $file->save();
                }
            }
        }

        return $val;

    }


    /**
     * @param string         $val
     * @param \DataContainer $dc
     * @return string
     */
    public function saveCopyrightCallback(string $val, \DataContainer $dc)
    {
        if ($dc instanceof DC_Folder) {
            $activeRecord = $dc->activeRecord;
            $folder       = FilesModel::findByUuid($activeRecord->uuid);
            $files        = FilesModel::findByPid($folder->uuid);

            if ($files !== null) {
                foreach ($files as $file) {
                    if ($file instanceof DC_Folder) {
                        continue;
                    }
                    if ($file->copyright === null || $file->copyright == '' || $file->copyright == $activeRecord->copyright) {
                        $file->copyright = $val;
                        $file->save();
                        continue;
                    }

                    $aCopyright = StringUtil::deserialize($file->copyright);
                    foreach ($aCopyright as $copyright) {
                        if ($copyright != '') {
                            return $val;
                        }
                    }
                    $file->copyright = $val;
                    $file->save();
                }
            }
        }

        return $val;
    }
}